### PR TITLE
Fix expand right-align when output rank > input rank

### LINKED ISSUE #53462: expand: wrong dimension alignment when output rank > input rank
Item 1 from master ticket

## Problem

`ttnn.expand` builds its repetition vector by comparing the input shape to the output shape **left-aligned**. PyTorch `expand` (and `ttnn.repeat`'s rank matching) use **right-alignment**.

When the output rank is higher than the input rank, the op can return **silent wrong data** or reject a valid expand with `TT_FATAL`.

## Example

Input shape `[3, 1]`, expand to `[3, 3, 1]`:

| | Result |
|--|--------|
| PyTorch | `[3, 3, 1]` (volume 27) |
| Current `ttnn.expand` | `[1, 9, 1]` (volume 9) — wrong |

Input shape `[2, 1]`, expand to `[4, 2, 3]`:

| | Result |
|--|--------|
| PyTorch | legal expand |
| Current `ttnn.expand` | `TT_FATAL` — should not fail |

## Location

`ttnn/cpp/ttnn/operations/data_movement/expand/expand.cpp` — `create_repetition_vector()`

## Suggested fix

Right-align when mapping input dims onto output dims: use `offset = new_rank - source_rank`, treat leading output dims as new size-1 axes, and compare trailing dims against the input shape.

## Test plan

Add unit tests in `tests/ttnn/unit_tests/base_functionality/test_expand.py` for rank-increasing expands (input rank < output rank), comparing against PyTorch.

### PR DESCRIPTION
**Pull request by VirdhatchaniKN**
Issue : https://github.com/tenstorrent/tt-metal/issues/53462

## Summary
- `create_repetition_vector()` in `expand.cpp` compared input dims to output dims **left-aligned**, producing wrong data or `TT_FATAL` for valid expands when output rank exceeded input rank.
- Fixed to **right-align** trailing dims (offset = new_rank − source_rank), matching `torch.expand` semantics and `repeat`'s own `match_input_rank` (`copy_backward`).

## Changes
| File | What |
|------|------|
| `ttnn/.../expand/expand.cpp` | Right-align dim mapping in `create_repetition_vector()` |
| `tests/.../test_expand.py` | 12 positive + 3 negative parametrised cases, shape assertions |

---

### Pipeline status

- [x]  Sanity
- [x]  (Runtime) Sanity Test
- [x]  Nightly for DM Category
- [x]  All model test
- [x]  Single card
- [x]  T3K
- [x]  Galaxy
- [x] (Tier 1) Models Device Perf Tests
- [x] ([TM-DM](https://github.com/tenstorrent/tt-metal/actions/runs/32479974955)) Data Movement Test Suite

---

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:virdhatchani/expand_53462)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:virdhatchani/expand_53462)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:virdhatchani/expand_53462)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:virdhatchani/expand_53462)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:virdhatchani/expand_53462)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:virdhatchani/expand_53462)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:virdhatchani/expand_53462)

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/data_movement/expand/expand.cpp [hunk 1]
```diff
@@ -15,15 +15,30 @@ ttsl::SmallVector<uint32_t> create_repetition_vector(const ttnn::Tensor& tensor,
...
+    // Right-align: input dims map to the trailing output dims, matching
+    // torch.expand semantics and repeat's own match_input_rank (copy_backward).
+    const auto offset = new_rank - source_rank;
+    for (size_t index = 0; index < new_rank; ++index) {
+        if (index < offset) {
+            TT_FATAL(
+                shape[index] >= 0,
+                "Leading dimension must be non-negative (got {}); it has no corresponding input dimension",
+                shape[index]);
             expansion_vector[index] = shape[index];
```
Comment (Aswinmcw) [inline ttnn/cpp/ttnn/operations/data_movement/expand/expand.cpp]: A new leading dim of size `1` produces a repetition of `1`, so if *every* entry ends up `1` the whole vector is all-ones and `ttnn::repeat` hits its early return at `repeat.cpp:394-397`, which returns `input_tensor` — not `working_tensor` — so the rank padding added by `match_input_rank` is thrown away.

`expand((3,1) -> (1,3,1))` gives repetition `{1,1,1}` and comes back as shape `(3,1)` instead of `(1,3,1)`. Before this PR that shape raised `TT_FATAL`; now it silently returns the wrong rank, which is the same failure class the issue is fixing. `(3,1) -> (1,1,3,1)` behaves the same.

Adding `[(3, 1), (1, 3, 1)]` to the new parametrize list should reproduce it against the shape assertion. Either return `working_tensor` from that early-return in `repeat`, or have `expand` reshape to the requested rank when `offset > 0`.

File: ttnn/cpp/ttnn/operations/data_movement/expand/expand.cpp [hunk 2]
```diff
@@ -15,15 +15,30 @@ ttsl::SmallVector<uint32_t> create_repetition_vector(const ttnn::Tensor& tensor,
...
         } else {
-            TT_FATAL(tensor_shape[index] == 1, "Only size 1 dimensions can be expanded in the output shape");
-            expansion_vector[index] = shape[index];
+            const auto src_dim = tensor_shape[static_cast<int32_t>(index - offset)];
+            if ((shape[index] == -1) || (shape[index] == static_cast<int32_t>(src_dim))) {
+                expansion_vector[index] = 1;
+            } else {
+                TT_FATAL(
+                    shape[index] >= 0,
+                    "Expand dimension size must be -1, match the input, or be non-negative (got {})",
```
Comment (Aswinmcw) [inline ttnn/cpp/ttnn/operations/data_movement/expand/expand.cpp]: This branch is only reached when `shape[index]` is negative and not `-1`, so "match the input, or be non-negative" reads contradictory next to the value it prints. `"Expand dimension size must be -1 or non-negative (got {})"` says the actual rule.

**Review by Aswinmcw — COMMENTED** [review]
One correctness gap below that the right-align change newly exposes; the other is a wording nit.

**Review by Aswinmcw — APPROVED** [review]
LGTM.

**Review by amorrisonTT — APPROVED** [review]

**Review by bbradelTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/data_movement/expand/expand.cpp [hunk 3]
```diff
@@ -7,26 +7,49 @@
...
-        if (index >= source_rank) {
+    TT_FATAL(source_rank <= new_rank, "Output rank must be >= input rank for expand");
+
+    ttsl::SmallVector<uint32_t> expansion_vector(new_rank);
+    ttsl::SmallVector<uint32_t> output_shape(new_rank);
+
+    // Right-align: input dims map to the trailing output dims, matching
+    // torch.expand semantics and repeat's own match_input_rank (copy_backward).
+    const auto offset = new_rank - source_rank;
+    for (size_t index = 0; index < new_rank; ++index) {
```
Comment (vtsilytskyiTT) [inline ttnn/cpp/ttnn/operations/data_movement/expand/expand.cpp]: If function right-aligns, it is natural to iterate backwards. Code should be simpler.

**Review by vtsilytskyiTT — APPROVED** [review]
Approved.

**Comment by VirdhatchaniKN** [comment]
https://github.com/tenstorrent/tt-metal/actions/runs/33028203004 - passed
